src/interaction/grab.rs

```diff
@@ -0,0 +1,139 @@
+        }
+    }
+}
+
+/// Attaches the grabbed entity to the player and marks the player occupied.
+fn handle_grab(
+    mut commands: Commands,
+    mut grab_state: ResMut<GrabState>,
+    mut events: EventReader<GrabEvent>,
+    player_q: Query<Entity, With<InteractionPlayer>>,
+) {
+    for GrabEvent(entity) in events.read() {
+        if grab_state.held.is_none() {
+            if let Ok(player) = player_q.single() {
+                commands.entity(player).add_child(*entity);
+                commands.entity(player).insert(Occupied);
+                grab_state.held = Some(*entity);
+            }
+        }
+    }
+}
+
+/// Detaches the held entity and places it in front of the player.
+fn handle_release(
+    mut commands: Commands,
+    mut grab_state: ResMut<GrabState>,
+    mut events: EventReader<ReleaseEvent>,
+    player_q: Query<Entity, With<InteractionPlayer>>,
+    player_tf_q: Query<&GlobalTransform, With<InteractionPlayer>>,
+    mut tf_q: Query<&mut Transform>,
+) {
+    // Release distance from player
+    const RELEASE_DISTANCE: f32 = 2.0;
+    for _ in events.read() {
+        if let Some(entity) = grab_state.held.take() {
+            if let Ok(player) = player_q.single() {
+                commands.entity(player).remove_children(&[entity]);
+                // Clear occupied tag
+                commands.entity(player).remove::<Occupied>();
+                if let (Ok(player_tf), Ok(mut item_tf)) =
+                    (player_tf_q.single(), tf_q.get_mut(entity))
+                {
+                    let forward = player_tf.forward();
+                    item_tf.translation = player_tf.translation()
+                        + forward * RELEASE_DISTANCE;
+                    item_tf.rotation = player_tf.rotation();
+                }
+            }
+        }
+    }
+}
+
+/// Ensure the held entity stays snapped on top of the player.
+fn update_snapping(
+    grab_state: Res<GrabState>,
+    mut tf_q: Query<&mut Transform>,
+) {
+    if let Some(entity) = grab_state.held {
+        if let Ok(mut item_tf) = tf_q.get_mut(entity) {
+            const HEIGHT_OFFSET: f32 = 1.5;
+            // Place item at player's head height
+            item_tf.translation = Vec3::Y * HEIGHT_OFFSET;
+
+            // Rotate the held item to match the player's rotation
+            item_tf.rotation = Quat::IDENTITY;
+        }
+    }
+}
+
+/// Marks an entity as grabbable.
+#[derive(Component, Reflect, Default)]
+#[reflect(Component)]
+pub struct Grabbable;
+
+/// Tags the player as occupied when holding an item.
+#[derive(Component, Reflect, Default)]
+#[reflect(Component)]
+pub struct Occupied;
+
+/// Tracks the currently held entity if any.
+#[derive(Resource, Default)]
+pub struct GrabState {
+    pub held: Option<Entity>,
+}
+
+/// Event to request grabbing a specified entity.
+#[derive(Event)]
+pub struct GrabEvent(pub Entity);
+
+/// Event to request releasing the currently held entity.
+#[derive(Event)]
+pub struct ReleaseEvent;
```

Comment:
You can use trigger for these, also, try to include who grab who and who release who ... if there are more than 1 player this logic will break ...

Reply:
Using trigger for this means you only need to embed the target item/object's `Entity` that you are grabbing, since you can fire a trigger event for a target entity (which could just be the player itself).